### ISSUE TITLE
Backbone with DCNv2 cannot implement training procedure

### ISSUE THREAD

**Issue by gsygsy96** [open-mmlab/mmdetection#8478]
I found that  models using resnet101 with DCNV2 cannot reproduce same gradient  after first backward with same config file, even with fix random seed and turn cudnn.deterministic==True.
Here is part of my config file:

> model = dict(type='DebugFrame',
>              debug = True,
>              backbone=dict(type='ResNet',
>                             depth=101,
>                             num_stages=4,
>                             out_indices=(0, 1, 2, 3),
>                             frozen_stages=0,
>                             norm_cfg=dict(type='BN2d', requires_grad=False),
>                             norm_eval=True,
>                             style='caffe',
>                             dcn=dict(type='DCNv2', deform_groups=1, fallback_on_stride=False),
>                             stage_with_dcn=(False, False, True, True)))
> optimizer = dict(type='SGD', lr=0.00005, momentum=0.)
> optimizer_config = dict()

Note that to debug this problem , I build a model called'DebugFrame', whose structure is shown below:

> @DETECTORS.register_module()
> class DebugFrame(BaseDetector):
>     def __init__(self,
>                  debug = True,
>                  backbone= None,
>                  *args,
>                  **kwargs):
>         super(DebugFrame, self).__init__()
>         self.debug = debug
>         self.backbone = build_backbone(backbone)
>         self.loss_type = torch.nn.L1Loss()
>         self.count = 0
> 
>     def init_weights(self):
>         pass
> 
>     def forward(self, return_loss=True, **kwargs):
> 
>         if return_loss:
>             return self.forward_train(**kwargs)
>         else:
>             return self.forward_test(**kwargs)
> 
>     def forward_train(self, img, *arg, **kwargs):
>         if self.debug:
>             save_dict = dict(backbone=dict(params=dict(),
>                                            grad=dict()))
> 
>             for k, v in self.backbone.named_parameters():
>                 save_dict['backbone']['params'][k] = v.clone().detach().cpu()
>                 if self.count == 0:
>                     save_dict['backbone']['grad'][k] = None
>                 else:
>                     save_dict['backbone']['grad'][k] = v.grad.clone().detach().cpu() if v.grad is not None else None
>             torch.save(save_dict, '/data1/CODE/detr_3d_sr/'
>                                   'work_dir/debug_backbone/backbone_%d.pth' % self.count)
>             print('BB saved')
>             self.count+=1
> 
>         img = rearrange(img, 'b n c h w -> (b n) c h w')
>         feats  = self.backbone(img)
>         losses = dict()
>         for i, f in enumerate(feats):
>             cur_gt = torch.ones_like(f)
>             losses[str(i)+'_loss'] = self.loss_type(cur_gt, f)*0.1
>         return losses
> 
> 
>     def aug_test(self, *args, **kwargs):
>         pass
> 
>     def extract_feat(self, *args, **kwargs):
>         pass
> 
>     def simple_test(self, *args, **kwargs):
>         pass

I checked gradient after the first backward(save_dict['backbone']['grad']) and the gradient varies firstly at layer 'layer4.2.conv1.weight', which receives gradient right after the most bottem DCN layer.

Samely, the gradient variation problem disappears when I shut down the DCN, with config '  stage_with_dcn=(False, False, False, False)))'

**Comment by gsygsy96** [comment]
can anyone reproduce this problem? Or it is just me who did something wrong?

**Comment by jbwang1997** [comment]
In some operations like interpolating, the randomness cannot be fixed because of the implementation of the cudnn.

**Comment by gsygsy96** [comment]
> In some operations like interpolating, the randomness cannot be fixed because of the implementation of the cudnn.

Thanks for replying! So the problem cannot be avoided right?
